### PR TITLE
battery: make time remaining estimation dependent on level flight cha…

### PR DESCRIPTION
**Pull request by KonradRudin**
…racteristis for FW

### Solved Problem
For fixed wing vehicles the remaining time estimate can significantly reduce when the vehicle is climbing for a longer period and trigger an RTL. The flight time estimate should be based on the level flight power consumption as this is the relevant data for the majority of the RTL flight profile.

Fixes #{Github issue ID}

### Solution
- Add simple level flight estimator in the fixed wing position controller. For now the estimator is only run when the vehicle controls the altitude or altitude rate, the altitude setpoint is constant and the vehicle is close to the altitude setpoint.
- The battery current_average is modified to only update the average, when level flight is detected for FW.

### Changelog Entry
For release notes:
```
Feature Battery remainig time estimate for FW is estimating time for level flight
```

### Alternatives
The current implementation makes the battery status dependent on the vehicle which is wrong. A better solution would be to actually have sort of a power consumption module which checks the battery that are relevant and make one consolidated estimate and battery overview. However, this would take considerable more effort and should be considered in the future.
Also the current implementation is dependent on the auto modes and does not update the estimate for manual modes.

### CODE REVIEW

File: src/lib/battery/battery.cpp [hunk 1]
```diff
@@ -307,20 +307,25 @@ float Battery::computeRemainingTime(float current_a)
 	float time_remaining_s = NAN;
 
 	if (_vehicle_status_sub.updated()) {
-		vehicle_status_s vehicle_status;
-
-		if (_vehicle_status_sub.copy(&vehicle_status)) {
-			_armed = (vehicle_status.arming_state == vehicle_status_s::ARMING_STATE_ARMED);
+		if (_vehicle_status_sub.update()) {
```
Comment (sfuhrer) [inline src/lib/battery/battery.cpp:310]: Why is this change needed? Looks out of context to me.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -2709,6 +2716,17 @@ FixedwingPositionControl::tecs_update_pitch_throttle(const float control_interva
 		     hgt_rate_sp);
 
 	tecs_status_publish(alt_sp, airspeed_sp, airspeed_rate_estimate, throttle_trim_adjusted);
+
+	// Estimate if we are in level flight.
+	// Tecs must be running, the height rate setpoint is not set or set to 0 and we are close to the target altitude and target altitude is not moving
+	if (_tecs_is_running && (!PX4_ISFINITE(hgt_rate_sp) || fabsf(hgt_rate_sp) < 0.1f)
+	    && fabsf(_current_altitude - alt_sp) < _param_nav_fw_alt_rad.get()
```
Comment (sfuhrer) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2723]: I think these thresholds are not ideal. 
E.g. they wouldn't classify the following flight as "in level flight", as the vehicle was oscillating around the altitude setpoint due to wind/thermal conditions. 
image

Could we instead look at tecs_status.altitude_reference? 
image

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 3]
```diff
@@ -2709,6 +2716,17 @@ FixedwingPositionControl::tecs_update_pitch_throttle(const float control_interva
...
+	    && fabsf((alt_sp - _last_tecs_alt_sp) / control_interval) < 0.1f) {
```
Comment (sfuhrer) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2724]: If we decide to keep these artificial thresholds: please make them constants.

**Review by sfuhrer — COMMENTED** [review]
The new message would need to be added to the logged_topics as well.

Could we make a more generic message that specifies with the states
CLIMB
LEVEL
DESCEND
? That could then also be used to set different places (eg to set flaps/spoilers). 

And it would be better to either not publish or set to a state NOT_ESTIMATED this new flight_level message if not estimated (if not in height rate controlled mode), instead of it getting published with the last state before switching out of a height rate controlled mode.

File: src/lib/battery/battery.cpp (hunk 1 above)
Comment (KonradRudin) [reply src/lib/battery/battery.cpp:310]: I need the vehicle_status later on since i need to check if it is a Fixedwing on the current estimation

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 2 above)
Comment (KonradRudin) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:2723]: It would be classified as level flight! alt_sp is constant, and i guess hgt_rate_sp is NaN (Be aware that i make the estimation in the fixedWingPositionControl and hgt_rate_sp is not tecs_status/height_rate_setpoint). The hgt_rate_sp is a feedforward term i can set to TECS (which in most cases except landing/starting is NaN i think) while height_rate_setpoint is the reference for the altitude rate controller.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 3 above)
Comment (KonradRudin) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:2724]: I will include that. Do you have any better value for the threshholds? I basically want to say that the feedforward height rate and the height setpoint changes should be around zero, hence i just chose 0.1. Open for more suggestions. I don't think this requires a parameter so i will make a constant out of it.

File: src/lib/battery/battery.cpp (hunk 1 above)
Comment (KonradRudin) [reply src/lib/battery/battery.cpp]: I reworked it a bit its closer again, but i need the vehicle_status outside that loop

**Comment by KonradRudin** [comment]
@sfuhrer I have included your comments. Also here a screenshot of the estimation which now estimates climb and descend as well during a mission.
Screenshot from 2023-11-23 15-30-54

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 4]
```diff
@@ -48,6 +48,8 @@ using matrix::Vector2d;
 using matrix::Vector3f;
 using matrix::wrap_pi;
 
+static constexpr float MAX_ALT_REF_RATE_FOR_LEVEL_FLIGHT{0.1f};
```
Comment (sfuhrer) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:51]: Please move to the header and use the convention of the FW position controller (kMax...)

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 2 above)
Comment (sfuhrer) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp]: ah right, my bad :+1:

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 5]
```diff
@@ -2709,6 +2718,32 @@ FixedwingPositionControl::tecs_update_pitch_throttle(const float control_interva
 		     hgt_rate_sp);
 
 	tecs_status_publish(alt_sp, airspeed_sp, airspeed_rate_estimate, throttle_trim_adjusted);
+
+	if (_tecs_is_running && !_vehicle_status.in_transition_mode
+	    && (_vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING)) {
+		// Check level flight: the height rate setpoint is not set or set to 0 and we are close to the target altitude and target altitude is not moving
+		if ((!PX4_ISFINITE(hgt_rate_sp) || fabsf(hgt_rate_sp) < MAX_ALT_REF_RATE_FOR_LEVEL_FLIGHT) &&
+		    fabsf(_current_altitude - alt_sp) < _param_nav_fw_alt_rad.get() &&
+		    fabsf((alt_sp - _last_tecs_alt_sp) / control_interval) < MAX_ALT_REF_RATE_FOR_LEVEL_FLIGHT) {
```
Comment (sfuhrer) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2727]: Ideally it would not be couple with the navigator acceptance logic, but I wouldn't have a good idea how to do it else.
You on purpose don't want to rely on `tecs_status.altitude_reference`?

**Review by sfuhrer — COMMENTED** [review]
Somehow the logging of flight_status_estimation is not working on my setup..could it be related to that you chose` uORB::PublicationData `instead of` uORB::Publication`? I'm not familiar with it, why did you chose it?

File: src/lib/battery/battery.cpp [hunk 6]
```diff
@@ -306,21 +306,28 @@ float Battery::computeRemainingTime(float current_a)
 {
 	float time_remaining_s = NAN;
 
-	if (_vehicle_status_sub.updated()) {
-		vehicle_status_s vehicle_status;
+	vehicle_status_s vehicle_status;
 
+	if (_vehicle_status_sub.updated()) {
```
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/lib/battery/battery.cpp]: fixed in 45d378b5f044e9a0e56e4e7daa47c16765874f0f

[the remark this answers is not in the document]
Comment (KonradRudin) [reply src/lib/battery/battery.cpp]: thanks

**Comment by KonradRudin** [comment]
> Somehow the logging of flight_status_estimation is not working on my setup..could it be related to that you chose`uORB::PublicationData`instead of` uORB::Publication`? I'm not familiar with it, why did you chose it?

That is strange, it worked for me. Publication data is not that different from publication, its just that i have not a proper single place where i could set the flight phase for all different use cases , so i need to store some information in the class. Its a bit cleaner this way instead of using another state variable.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 5 above)
Comment (KonradRudin) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:2727]: That is the internally filtered altitude reference i chose the setpoint since when you command a new one, it immediately switches to the climb flight phase and not after the filtered one ramped up. So the flight phase is more coupled to the commands not not the internal filter. oth would be viable though.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 4 above)
Comment (KonradRudin) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp]: Moved in https://github.com/PX4/PX4-Autopilot/pull/22401/commits/d182e0dd742a988587540a592448559c5dc0d437

**Comment by KonradRudin** [comment]
rebased on main

**Review by sfuhrer — DISMISSED** [review]
Look good for me now, let's get some real world experience and tune the threshold if needed.

**Comment by KonradRudin** [comment]
rebased on main

**Comment by KonradRudin** [comment]
Hey @sfuhrer we have finally gathered some logs for it. See here:
Screenshot from 2024-02-05 15-57-25
Screenshot from 2024-02-05 16-11-55

Two things to note here. It takes the battery estimation a long time to recover from the transition (up to 4-5 min) and we estimate the level flight phase early on a change (due to accepting as level flight if the setpoint does not change, and we are within the acceptance radius, which can be large).
So to move on, potential action: 

- should we reset the filter when we change the modes? We would have jumps in the time estimation then, but would have faster convergence after transition. 
- Should we reformulate the level flight acceptance `fabsf(_current_altitude - tecs_output.altitude_reference) < _param_nav_fw_alt_rad.get()` ? Either have a lower bound or reformulate that the current altitude rate must be below a threshold (although that threshold would also depend on the control tuning and outside conditions, potentially need to be a parameter then.

**Review by sfuhrer — APPROVED** [review]
Looks like in a good state to bring in. It's a trade off between having a robust "level flight" identifier and not adding more settings than necessary. This is anyway a good iteration and may be good enough long term as well.

File: src/lib/battery/battery.cpp [hunk 7]
```diff
@@ -311,16 +311,23 @@ float Battery::computeRemainingTime(float current_a)
...
 	}
 
 	if (_armed && PX4_ISFINITE(current_a)) {
-		// only update with positive numbers
-		_current_average_filter_a.update(fmaxf(current_a, 0.f));
+		// For FW only update when we are in level flight
+		if (!_vehicle_status_is_fw || ((hrt_absolute_time() - _flight_phase_estimation_sub.get().timestamp) < 2_s
+					       && _flight_phase_estimation_sub.get().flight_phase == flight_phase_estimation_s::FLIGHT_PHASE_LEVEL)) {
+			// only update with positive numbers
+			_current_average_filter_a.update(fmaxf(current_a, 0.f));
```
Comment (MaEtUgR) [inline src/lib/battery/battery.cpp:329]: @KonradRudin Can't you just always clamp to consider current consumption during flight instead of adding so much complexity to in some rare cases also consider charging the battery? Is charging the battery in air a thing?

Comment (MaEtUgR) [reply src/lib/battery/battery.cpp:329]: Nevermind, I read it wrongly.
